src/main/java/org/spongepowered/api/event/cause/EventContext.java

```diff
@@ -0,0 +1,182 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.event.cause;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import org.spongepowered.api.util.ResettableBuilder;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.StringJoiner;
+
+import javax.annotation.Nullable;
+
+/**
+ * Provides context for an event outside of the direct chain of causes present
+ * in the event's {@link Cause}.
+ */
+public final class EventContext {
+
+    private static final EventContext EMPTY_CONTEXT = new EventContext(ImmutableMap.of());
+
+    /**
+     * Gets an empty context.
+     * 
+     * @return The empty context
+     */
+    public static EventContext empty() {
+        return EMPTY_CONTEXT;
+    }
+
+    /**
+     * Creates a new {@link EventContext} from the given map of entries.
+     * 
+     * @param entries The context entries
+     * @return The new EventContext
+     */
+    public static EventContext of(Map<EventContextKey<?>, Object> entries) {
+        checkNotNull(entries, "Context entries cannot be null");
+        Map<String, Object> ctx = new HashMap<>(); 
+        for (Map.Entry<EventContextKey<?>, Object> entry : entries.entrySet()) {
+            checkNotNull(entry.getKey(), "Entries cannot contain null keys");
+            checkNotNull(entry.getValue(), "Entries cannot contain null values");
+            ctx.put(entry.getKey().getId(), entry.getValue());
+        }
+        return new EventContext(ctx);
+    }
+
+    /**
+     * Creates a new builder for creating an {@link EventContext}.
+     * 
+     * @return The new builder
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    private final Map<String, Object> entries;
+
+    EventContext(Map<String, Object> values) {
+        this.entries = ImmutableMap.copyOf(values);
+    }
+
+    /**
+     * Gets the value corresponding to the given key from the context.
+     * 
+     * @param key The key
+     * @return The context value, if found
+     */
+    @SuppressWarnings("unchecked")
+    public <T> Optional<T> get(EventContextKey<T> key) {
```

Comment:
A `<T> T require(EventContextKey<T> key)` method (similar to https://github.com/SpongePowered/SpongeAPI/commit/4a82ccf8ed7c558296e203ceece571794b735c98) would be great here.